ethereum/eip712/eip712_test.go

```diff
@@ -428,7 +519,231 @@ func (suite *EIP712TestSuite) verifyEIP712SignatureVerification(expectedSuccess
 	randBytes := make([]byte, len(signBytes))
 	copy(randBytes, signBytes)
 	// Change the first element of signBytes to a different value
-	randBytes[0] = (signBytes[0] + 10) % 128
+	randBytes[0] = (signBytes[0] + 10) % 255
 	res = pubKey.VerifySignature(randBytes, sig)
 	suite.Require().False(res)
 }
+
+// verifySignDocFlattening tests the flattening algorithm against the sign doc's JSON payload,
+// using verifyPayloadAgainstFlattened.
+func (suite *EIP712TestSuite) verifySignDocFlattening(signDoc []byte) {
+	payload := gjson.ParseBytes(signDoc)
+	suite.Require().True(payload.IsObject())
+
+	flattened, _, err := eip712.FlattenPayloadMessages(payload)
+	suite.Require().NoError(err)
+
+	suite.verifyPayloadAgainstFlattened(payload, flattened)
+}
+
+// verifyPayloadAgainstFlattened compares a payload against its flattened counterpart to ensure that
+// the flattening algorithm behaved as expected.
+func (suite *EIP712TestSuite) verifyPayloadAgainstFlattened(payload gjson.Result, flattened gjson.Result) {
+	payloadMap, ok := payload.Value().(map[string]interface{})
+	suite.Require().True(ok)
+	flattenedMap, ok := flattened.Value().(map[string]interface{})
+	suite.Require().True(ok)
+
+	suite.verifyPayloadMapAgainstFlattenedMap(payloadMap, flattenedMap)
+}
+
+// verifyPayloadMapAgainstFlattenedMap directly compares two JSON maps in Go representations to
+// test flattening.
+func (suite *EIP712TestSuite) verifyPayloadMapAgainstFlattenedMap(original map[string]interface{}, flattened map[string]interface{}) {
+	interfaceMessages, ok := original["msgs"]
+	suite.Require().True(ok)
+
+	messages, ok := interfaceMessages.([]interface{})
+	suite.Require().True(ok)
+
+	// Verify message contents
+	for i, msg := range messages {
+		flattenedMsg, ok := flattened[fmt.Sprintf("msg%d", i)]
+		suite.Require().True(ok)
+
+		flattenedMsgJSON, ok := flattenedMsg.(map[string]interface{})
+		suite.Require().True(ok)
+
+		suite.Require().True(reflect.DeepEqual(flattenedMsgJSON, msg))
+	}
+
+	// Verify new payload does not have msgs field
+	_, ok = flattened["msgs"]
+	suite.Require().False(ok)
+
+	// Verify number of total keys
+	numKeysOriginal := len(original)
+	numKeysFlattened := len(flattened)
+	numMessages := len(messages)
+
+	// + N keys, then -1 for msgs
+	suite.Require().Equal(numKeysFlattened, numKeysOriginal+numMessages-1)
+
+	// Verify contents of remaining keys
+	for k, obj := range original {
+		if k == "msgs" {
+			continue
+		}
+
+		flattenedObj, ok := flattened[k]
+		suite.Require().True(ok)
+
+		suite.Require().Equal(obj, flattenedObj)
+		suite.Require().True(reflect.DeepEqual(obj, flattenedObj))
```

Comment:
The `Equal()` func already uses the `reflect.DeepEqual`. [Here's the code](https://github.com/stretchr/testify/blob/0ab3ce1249292a7221058b9e370472bca8f04813/assert/assertions.go#L327-L353), it uses the [ObjectsAreEqual](https://github.com/stretchr/testify/blob/0ab3ce1249292a7221058b9e370472bca8f04813/assert/assertions.go#L55-L76) func which uses the reflect pkg
```suggestion
		suite.Require().Equal(obj, flattenedObj)
```